File: langgraph-typescript/src/tools.ts
import fs from "node:fs";
import path from "node:path";
import { search } from "duck-duck-scrape";
import { sanitizeMermaidInMarkdown, sanitizeMermaidSource } from "../../shared/mermaid-sanitize.js";

export class Sandbox {
  constructor(
    public readonly root: string,
    public readonly knowledgeDir: string,
  ) {
    fs.mkdirSync(path.join(root, "src"), { recursive: true });
    fs.mkdirSync(path.join(root, "diagrams"), { recursive: true });
    fs.mkdirSync(path.join(root, "assets"), { recursive: true });
  }

  private safePath(relative: string): string {
    const cleaned = relative.replace(/\\/g, "/").replace(/^\/+/, "");
    if (cleaned.split("/").includes("..")) {
      throw new Error("Path traversal is not allowed");
    }
    const full = path.resolve(this.root, cleaned);
    if (!full.startsWith(path.resolve(this.root))) {
      throw new Error("Path escapes sandbox");
    }
    return full;
  }

  writeFile(rel: string, content: string): string {
    if (content.includes("```mermaid")) {
      content = sanitizeMermaidInMarkdown(content);
    }
    const target = this.safePath(rel);
    fs.mkdirSync(path.dirname(target), { recursive: true });
    fs.writeFileSync(target, content, "utf8");
    return `Wrote ${rel} (${content.length} chars)`;
  }

  readFile(rel: string): string {
    const target = this.safePath(rel);
    if (!fs.existsSync(target)) return `ERROR: file not found: ${rel}`;
    return fs.readFileSync(target, "utf8");
  }

  listFiles(prefix = ""): string {
    const base = prefix ? this.safePath(prefix) : this.root;
    if (!fs.existsSync(base)) return "(empty)";
    const files: string[] = [];
    const walk = (dir: string) => {
      for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
        const full = path.join(dir, entry.name);
        if (entry.isDirectory()) walk(full);
        else files.push(path.relative(this.root, full));
      }
    };
    walk(base);
    files.sort();
    return files.length ? files.join("\n") : "(empty)";
  }

  searchKnowledge(query: string, limit = 3): string {
    const tokens = query
      .toLowerCase()
      .split(/\W+/)
      .filter((t) => t.length > 2);
    const scored: { score: number; name: string; text: string }[] = [];
    for (const name of fs.readdirSync(this.knowledgeDir)) {
      if (!name.endsWith(".md")) continue;
      const text = fs.readFileSync(path.join(this.knowledgeDir, name), "utf8");
      const lower = text.toLowerCase();
      const score = tokens.length
        ? tokens.reduce((acc, tok) => acc + (lower.split(tok).length - 1), 0)
        : 1;
      if (score > 0) scored.push({ score, name, text });
    }
    scored.sort((a, b) => b.score - a.score);
    if (!scored.length) return "No knowledge matches.";
    return scored
      .slice(0, limit)
      .map((s) => `### ${s.name} (score=${s.score})\n${s.text.slice(0, 2500)}`)
      .join("\n\n");
  }

  async webSearch(query: string, limit = 5): Promise<string> {
    try {
      const results = await search(query, { safeSearch: 0 });
      const items = (results.results ?? []).slice(0, limit);
      if (!items.length) return "No web results found.";
      return items
        .map((item) => {
          const title = item.title ?? "No title";
          const href = item.url ?? "";
          const body = (item.description ?? "").slice(0, 500);
          return `- **${title}**\n  ${href}\n  ${body}`;
        })
        .join("\n\n");
    } catch (err) {
      return `Web search unavailable: ${err}. Use search_knowledge as fallback.`;
    }
  }

  writeMermaid(rel: string, content: string): string {
    if (!content.trim()) throw new Error("Mermaid content cannot be empty");
    let cleaned = rel.replace(/\\/g, "/").replace(/^\/+/, "");
    if (!cleaned.endsWith(".mmd")) cleaned = `${cleaned}.mmd`;
    if (!cleaned.startsWith("diagrams/")) cleaned = `diagrams/${cleaned.replace(/^\//, "")}`;
    let body = content.trim();
    if (body.startsWith("```")) {
      body = body.replace(/^```(?:mermaid)?\s*/, "").replace(/\s*```$/, "");
    }
    body = sanitizeMermaidSource(body);
    return this.writeFile(cleaned, body + "\n");
  }

  async generateImage(prompt: string, rel: string): Promise<string> {
    const apiKey = process.env.OPENAI_API_KEY?.trim();
    if (!apiKey) {
      return (
        "ERROR: OPENAI_API_KEY required for generate_image " +
        "(illustrator needs OpenAI even when LLM uses DeepSeek)."
      );
    }
    let target = rel.replace(/\\/g, "/").replace(/^\/+/, "");
    if (!target.startsWith("assets/")) target = `assets/${target.replace(/^\//, "")}`;
    if (!target.toLowerCase().endsWith(".png")) target = `${target}.png`;
    try {
      const response = await fetch("https://api.openai.com/v1/images/generations", {
        method: "POST",
        headers: {
          Authorization: `Bearer ${apiKey}`,
          "Content-Type": "application/json",
        },
        // gpt-image models reject response_format; they return b64_json by default.
        body: JSON.stringify({
          model: "gpt-image-1",
          prompt,
          size: "1024x1024",
          n: 1,
        }),
      });
      if (!response.ok) {
        const text = await response.text();
        return `ERROR generating image: ${response.status} ${text}`;
      }
      const data = (await response.json()) as {
        data?: { b64_json?: string; url?: string }[];
      };
      const item = data.data?.[0];
      let bytes: Buffer | undefined;
      if (item?.b64_json) {
        bytes = Buffer.from(item.b64_json, "base64");
      } else if (item?.url) {
        const imgRes = await fetch(item.url);
        if (!imgRes.ok) {
          return `ERROR generating image: failed to download url (${imgRes.status})`;
        }
        bytes = Buffer.from(await imgRes.arrayBuffer());
      }
      if (!bytes) return "ERROR generating image: empty response";
      const full = this.safePath(target);
      fs.mkdirSync(path.dirname(full), { recursive: true });
      fs.writeFileSync(full, bytes);
      return `Generated image at ${target}`;
    } catch (err) {
      return `ERROR generating image: ${err}`;
    }
  }

  listWrittenFiles(): string[] {
    const files: string[] = [];
    const walk = (dir: string) => {
      for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
        const full = path.join(dir, entry.name);
        if (entry.isDirectory()) walk(full);
        else if (entry.name !== "summary.json") {
          files.push(path.relative(this.root, full));
        }
      }
    };
    walk(this.root);
    return files.sort();
  }
}

File: langflow-python/app/tools.py
"""Sandbox filesystem tools + local knowledge search + external capabilities."""

from __future__ import annotations

import base64
import os
import re
from pathlib import Path

from feature_delivery_api.mermaid import sanitize_mermaid_in_markdown, sanitize_mermaid_source


class Sandbox:
    def __init__(self, root: Path, knowledge_dir: Path) -> None:
        self.root = root
        self.knowledge_dir = knowledge_dir
        self.root.mkdir(parents=True, exist_ok=True)
        (self.root / "src").mkdir(exist_ok=True)
        (self.root / "diagrams").mkdir(exist_ok=True)
        (self.root / "assets").mkdir(exist_ok=True)

    def _safe_path(self, relative: str) -> Path:
        relative = relative.lstrip("/").replace("\\", "/")
        if ".." in Path(relative).parts:
            raise ValueError("Path traversal is not allowed")
        path = (self.root / relative).resolve()
        if not str(path).startswith(str(self.root.resolve())):
            raise ValueError("Path escapes sandbox")
        return path

    def write_file(self, path: str, content: str) -> str:
        if "```mermaid" in content:
            content = sanitize_mermaid_in_markdown(content)
        target = self._safe_path(path)
        target.parent.mkdir(parents=True, exist_ok=True)
        target.write_text(content, encoding="utf-8")
        return f"Wrote {path} ({len(content)} chars)"

    def read_file(self, path: str) -> str:
        target = self._safe_path(path)
        if not target.exists():
            return f"ERROR: file not found: {path}"
        return target.read_text(encoding="utf-8")

    def list_files(self, prefix: str = "") -> str:
        base = self._safe_path(prefix) if prefix else self.root
        if not base.exists():
            return "(empty)"
        files: list[str] = []
        for p in sorted(base.rglob("*")):
            if p.is_file():
                files.append(str(p.relative_to(self.root)))
        return "\n".join(files) if files else "(empty)"

    def search_knowledge(self, query: str, limit: int = 3) -> str:
        tokens = [t for t in re.split(r"\W+", query.lower()) if len(t) > 2]
        scored: list[tuple[int, Path, str]] = []
        for path in self.knowledge_dir.glob("*.md"):
            text = path.read_text(encoding="utf-8")
            lower = text.lower()
            score = sum(lower.count(tok) for tok in tokens) if tokens else 1
            if score > 0:
                scored.append((score, path, text))
        scored.sort(key=lambda x: x[0], reverse=True)
        if not scored:
            return "No knowledge matches."
        chunks: list[str] = []
        for score, path, text in scored[:limit]:
            chunks.append(f"### {path.name} (score={score})\n{text[:2500]}")
        return "\n\n".join(chunks)

    def web_search(self, query: str, limit: int = 5) -> str:
        try:
            try:
                from ddgs import DDGS
            except ImportError:
                from duckduckgo_search import DDGS  # noqa: F401 — legacy fallback

            with DDGS() as ddgs:
                results = list(ddgs.text(query, max_results=limit))
            if not results:
                return "No web results found."
            chunks: list[str] = []
            for item in results:
                title = item.get("title", "No title")
                href = item.get("href", "")
                body = (item.get("body") or "")[:500]
                chunks.append(f"- **{title}**\n  {href}\n  {body}")
            return "\n\n".join(chunks)
        except Exception as exc:
            return (
                f"Web search unavailable: {exc}. "
                "Use search_knowledge as fallback."
            )

    def write_mermaid(self, path: str, content: str) -> str:
        if not content.strip():
            raise ValueError("Mermaid content cannot be empty")
        rel = path.lstrip("/").replace("\\", "/")
        if not rel.endswith(".mmd"):
            rel = f"{rel}.mmd"
        if not rel.startswith("diagrams/"):
            rel = f"diagrams/{rel.lstrip('/')}"
        cleaned = content.strip()
        if cleaned.startswith("```"):
            cleaned = re.sub(r"^```(?:mermaid)?\s*", "", cleaned)
            cleaned = re.sub(r"\s*```$", "", cleaned)
        cleaned = sanitize_mermaid_source(cleaned)
        return self.write_file(rel, cleaned + "\n")

    def generate_image(self, prompt: str, path: str) -> str:
        api_key = os.environ.get("OPENAI_API_KEY", "").strip()
        if not api_key:
            return (
                "ERROR: OPENAI_API_KEY required for generate_image "
                "(illustrator needs OpenAI even when LLM uses DeepSeek)."
            )
        rel = path.lstrip("/").replace("\\", "/")
        if not rel.startswith("assets/"):
            rel = f"assets/{rel.lstrip('/')}"
        if not rel.lower().endswith(".png"):
            rel = f"{rel}.png"
        try:
            from openai import OpenAI

            client = OpenAI(api_key=api_key)
            response = client.images.generate(
                model="gpt-image-1",
                prompt=prompt,
                size="1024x1024",
                n=1,
            )
            item = response.data[0]
            if getattr(item, "b64_json", None):
                image_data = base64.b64decode(item.b64_json)
            elif getattr(item, "url", None):
                import urllib.request

                with urllib.request.urlopen(item.url) as resp:
                    image_data = resp.read()
            else:
                return "ERROR generating image: empty response"
            target = self._safe_path(rel)
            target.parent.mkdir(parents=True, exist_ok=True)
            target.write_bytes(image_data)
            return f"Generated image at {rel}"
        except Exception as exc:
            return f"ERROR generating image: {exc}"

    def list_written_files(self) -> list[str]:
        files: list[str] = []
        for p in sorted(self.root.rglob("*")):
            if p.is_file() and p.name != "summary.json":
                files.append(str(p.relative_to(self.root)))
        return files
